apps/src/blockly/addons/cdoFieldPicker.js

```diff
@@ -0,0 +1,129 @@
+import GoogleBlockly from 'blockly/core';
+
+const BUTTON_CORNER_RADIUS = 3;
+const BUTTON_INNER_HEIGHT = 16;
+import {DEFAULT_SOUND} from '@cdo/apps/blockly/constants';
+
+/**
+ * This is a customized field which the user clicks to select an option from a customized picker,
+ * for example, the location of a sprite from a grid or a sound file from a customized modal.
+ * @param value The initial value of the field.
+ * @param onChange The function that handles the field's editor.
+ * @param onDisplay The function tht handles how the field text is displayed.
+ * @param buttonIcon If the field displyas a button, this is the string that is displayed on the button.
+ */
+export default class CdoFieldPicker extends GoogleBlockly.Field {
+  constructor(value, onChange, onDisplay, buttonIcon) {
+    super(value);
+    this.onChange = onChange;
+    this.onDisplay = onDisplay;
+    this.buttonIcon = buttonIcon;
+
+    this.SERIALIZABLE = true;
+  }
+
+  static fromJson(options) {
+    return new CdoFieldPicker(
+      options.value,
+      options.onChange,
+      options.onDisplay,
+      options.buttonIcon
+    );
+  }
+
+  /**
+   * Validate the changes to a field's value before they are set.
+   * @override
+   */
+  doClassValidation_(newValue) {
+    if (typeof newValue !== 'string') {
+      return null;
+    }
+    // Handle 'play sound' block with default param from CDO blockly.
+    // TODO: Remove when sprite lab is migrated to Google blockly.
+    if (newValue === 'Choose') {
+      return DEFAULT_SOUND;
```

Comment:
This feels too specific to be in the class that is used to make the fields for location blocks. I wonder if each time we make this kind of field if we should pass its value validator in as a function to the constructor.